sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/JoinEstimation.scala

```diff
@@ -0,0 +1,175 @@
+              newMin1, newMax1))
+            intersectedStats.put(rightKey.name, intersectedColumnStat(rightKeyStats, minNdv,
+              newMin2, newMax2))
+          } else {
+            // Set ndvDenom to zero to indicate that this join should have no output
+            ndvDenom = 0
+          }
+        }
+        i += 1
+      }
+
+      // 2. Estimate the number of output rows
+      val leftRows = leftStats.rowCount.get
+      val rightRows = rightStats.rowCount.get
+      val outputRows: BigInt = if (ndvDenom == 0) {
+        // One of the join key pairs is disjoint, thus the two sides of join is disjoint.
+        0
+      } else if (ndvDenom < 0) {
+        // There isn't join keys or column stats for any of the join key pairs, we estimate like
+        // cartesian product.
+        leftRows * rightRows
+      } else {
+        ceil(BigDecimal(leftRows * rightRows) / BigDecimal(ndvDenom))
+      }
+
+      // 3. Update statistics based on the output of join
+      val originalStats = leftStats.colStats ++ rightStats.colStats
+      val outputWithStat = join.output.filter(attr => originalStats.contains(attr.name))
+
+      val outputColStats = new mutable.HashMap[String, ColumnStat]()
+      if (ndvDenom == 0) {
+        // empty output
+        outputWithStat.foreach(a => outputColStats.put(a.name, emptyColumnStat(a.dataType)))
+      } else if (ndvDenom < 0) {
+        // cartesian product, column stats will not change
+        outputWithStat.foreach(a => outputColStats.put(a.name, originalStats(a.name)))
+      } else {
+        val leftRatio = BigDecimal(outputRows) / BigDecimal(leftRows)
+        val rightRatio = BigDecimal(outputRows) / BigDecimal(rightRows)
+        outputWithStat.foreach { a =>
+          // check if this attribute is a join key
+          if (intersectedStats.contains(a.name)) {
+            outputColStats.put(a.name, intersectedStats(a.name))
+          } else {
+            val oldColStat = originalStats(a.name)
+            val oldNdv = oldColStat.distinctCount
+            // We only change (scale down) the number of distinct values if the number of rows
+            // decreases after join, because join won't produce new values even if the number of
+            // rows increases.
+            val newNdv = if (left.outputSet.contains(a) && leftRatio < 1) {
+              ceil(BigDecimal(oldNdv) * leftRatio)
+            } else if (right.outputSet.contains(a) && rightRatio < 1) {
+              ceil(BigDecimal(oldNdv) * rightRatio)
+            } else {
+              oldNdv
+            }
+            outputColStats.put(a.name, oldColStat.copy(distinctCount = newNdv, nullCount = 0))
+          }
+        }
+      }
+
+      Some(Statistics(
+        sizeInBytes = outputRows * getRowSize(join.output, outputColStats.toMap),
+        rowCount = Some(outputRows),
+        colStats = outputColStats.toMap,
+        isBroadcastable = false))
+
+    case _ => None
+  }
```

Comment:
Thanks for the advice. We will consider to have such function when we check in the code for Filter estimation, then we can look at this more comprehensively. I think the pr will be submitted soon.